File: shopreel/sources/amazon.py
# -*- coding: utf-8 -*-
"""Amazon Product Advertising API 5.0 (SearchItems).

필요 환경변수
  AMAZON_ACCESS_KEY / AMAZON_SECRET_KEY / AMAZON_ASSOC_TAG
선택
  AMAZON_HOST(기본 webservices.amazon.com), AMAZON_REGION(기본 us-east-1),
  AMAZON_MARKETPLACE(기본 www.amazon.com), AMAZON_KEYWORDS, AMAZON_BROWSE_NODE
"""

from __future__ import annotations

import datetime
import hashlib
import hmac
import json
from typing import Dict, List

from ..config import Config
from ..models import Product
from .base import Source, env, http, to_float, to_int

SERVICE = "ProductAdvertisingAPI"
PATH = "/paapi5/searchitems"
TARGET = "com.amazon.paapi5.v1.ProductAdvertisingAPIv1.SearchItems"
RESOURCES = [
    "Images.Primary.Large", "ItemInfo.Title", "ItemInfo.Features",
    "ItemInfo.ByLineInfo", "Offers.Listings.Price",
    "Offers.Listings.SavingBasis", "CustomerReviews.StarRating",
    "CustomerReviews.Count", "BrowseNodeInfo.BrowseNodes",
]


def _sign(key: bytes, msg: str) -> bytes:
    return hmac.new(key, msg.encode("utf-8"), hashlib.sha256).digest()


def sigv4_headers(host: str, region: str, access: str, secret: str, body: str) -> Dict[str, str]:
    """AWS SigV4 서명 헤더."""
    now = datetime.datetime.now(datetime.timezone.utc)
    amz_date = now.strftime("%Y%m%dT%H%M%SZ")
    stamp = now.strftime("%Y%m%d")

    headers = {
        "content-encoding": "amz-1.0",
        "content-type": "application/json; charset=utf-8",
        "host": host,
        "x-amz-date": amz_date,
        "x-amz-target": TARGET,
    }
    signed = ";".join(sorted(headers))
    canonical_headers = "".join(f"{k}:{headers[k]}\n" for k in sorted(headers))
    payload_hash = hashlib.sha256(body.encode("utf-8")).hexdigest()
    canonical = "\n".join(["POST", PATH, "", canonical_headers, signed, payload_hash])

    scope = f"{stamp}/{region}/{SERVICE}/aws4_request"
    to_sign = "\n".join(["AWS4-HMAC-SHA256", amz_date, scope,
                         hashlib.sha256(canonical.encode("utf-8")).hexdigest()])
    k = _sign(f"AWS4{secret}".encode("utf-8"), stamp)
    k = _sign(k, region)
    k = _sign(k, SERVICE)
    k = _sign(k, "aws4_request")
    signature = hmac.new(k, to_sign.encode("utf-8"), hashlib.sha256).hexdigest()

    headers["Authorization"] = (
        f"AWS4-HMAC-SHA256 Credential={access}/{scope}, "
        f"SignedHeaders={signed}, Signature={signature}")
    return headers


class AmazonSource(Source):
    name = "amazon"
    network = "amazon"
    needs = ("AMAZON_ACCESS_KEY", "AMAZON_SECRET_KEY", "AMAZON_ASSOC_TAG")

    def fetch(self, cfg: Config, limit: int) -> List[Product]:
        host = env("AMAZON_HOST") or "webservices.amazon.com"
        region = env("AMAZON_REGION") or "us-east-1"
        payload = {
            "Keywords": env("AMAZON_KEYWORDS") or "best sellers",
            "PartnerTag": env("AMAZON_ASSOC_TAG"),
            "PartnerType": "Associates",
            "Marketplace": env("AMAZON_MARKETPLACE") or "www.amazon.com",
            "ItemCount": min(10, max(1, limit)),      # PA-API 는 호출당 최대 10개
            "SortBy": "Relevance",
            "Resources": RESOURCES,
        }
        node = env("AMAZON_BROWSE_NODE")
        if node:
            payload["BrowseNodeId"] = node
        body = json.dumps(payload)
        headers = sigv4_headers(host, region, env("AMAZON_ACCESS_KEY"),
                                env("AMAZON_SECRET_KEY"), body)
        raw = http(f"https://{host}{PATH}", method="POST", headers=headers,
                   data=body.encode("utf-8"))
        data = json.loads(raw.decode("utf-8", "replace"))

        out: List[Product] = []
        for item in (data.get("SearchResult", {}).get("Items", []) or [])[:limit]:
            info = item.get("ItemInfo", {}) or {}
            listing = ((item.get("Offers", {}) or {}).get("Listings") or [{}])[0]
            price = (listing.get("Price") or {})
            saving = (listing.get("SavingBasis") or {})
            reviews = item.get("CustomerReviews", {}) or {}
            nodes = (item.get("BrowseNodeInfo", {}) or {}).get("BrowseNodes") or [{}]
            features = ((info.get("Features") or {}).get("DisplayValues") or [])
            out.append(self.product(
                product_id=str(item.get("ASIN", "")),
                title=str(((info.get("Title") or {}).get("DisplayValue")) or ""),
                url=str(item.get("DetailPageURL", "")),
                image_url=str(((item.get("Images", {}) or {}).get("Primary", {})
                               .get("Large", {}) or {}).get("URL", "")),
                price=to_float(price.get("Amount")),
                orig_price=to_float(saving.get("Amount")),
                currency=str(price.get("Currency") or "USD"),
                rating=to_float((reviews.get("StarRating") or {}).get("Value")),
                reviews=to_int((reviews.get("Count") or {}).get("Value")),
                category=str(nodes[0].get("DisplayName", "")),
                shop=str(((info.get("ByLineInfo") or {}).get("Brand") or {})
                         .get("DisplayValue", "")),
                highlights=[str(f) for f in features[:3]],
                raw=item,
            ))
        return out

File: server/feeds/amazon.js
/**
 * Amazon Product Advertising API 5.0 어댑터.
 *
 * ⚠ 가입 문턱이 있다: 어소시에이트 계정을 만든 뒤 **180일 안에 적격 판매 3건**이 있어야
 *    PA-API 키가 열린다. 그래서 처음에는 국내 판매처부터 붙이고, 나중에 여는 걸 권한다.
 *
 * 인증: AWS SigV4 (service = ProductAdvertisingAPI).
 */
import crypto from 'node:crypto';

const REGION = process.env.AMAZON_REGION || 'us-east-1';
const HOST = process.env.AMAZON_HOST || 'webservices.amazon.com';
const PATH = '/paapi5/searchitems';
const SERVICE = 'ProductAdvertisingAPI';

export const id = 'amazon';
export const name = 'Amazon';
export const limits = { callsPerSecond: 1, itemsPerCall: 10 };

export function isConfigured() {
  return Boolean(process.env.AMAZON_ACCESS_KEY && process.env.AMAZON_SECRET_KEY && process.env.AMAZON_ASSOC_TAG);
}

const sha256 = (s) => crypto.createHash('sha256').update(s, 'utf8').digest('hex');
const hmac = (key, s) => crypto.createHmac('sha256', key).update(s, 'utf8').digest();

/** AWS SigV4 서명 헤더 생성 */
function signedHeaders(payload, target) {
  const now = new Date();
  const amzDate = now.toISOString().replace(/[:-]|\.\d{3}/g, '');
  const dateStamp = amzDate.slice(0, 8);

  const canonicalHeaders =
    `content-encoding:amz-1.0\n` +
    `host:${HOST}\n` +
    `x-amz-date:${amzDate}\n` +
    `x-amz-target:${target}\n`;
  const signedHeaderNames = 'content-encoding;host;x-amz-date;x-amz-target';
  const canonicalRequest = ['POST', PATH, '', canonicalHeaders, signedHeaderNames, sha256(payload)].join('\n');

  const scope = `${dateStamp}/${REGION}/${SERVICE}/aws4_request`;
  const stringToSign = ['AWS4-HMAC-SHA256', amzDate, scope, sha256(canonicalRequest)].join('\n');

  let key = hmac(`AWS4${process.env.AMAZON_SECRET_KEY}`, dateStamp);
  key = hmac(key, REGION);
  key = hmac(key, SERVICE);
  key = hmac(key, 'aws4_request');
  const signature = crypto.createHmac('sha256', key).update(stringToSign, 'utf8').digest('hex');

  return {
    'content-encoding': 'amz-1.0',
    'content-type': 'application/json; charset=utf-8',
    host: HOST,
    'x-amz-date': amzDate,
    'x-amz-target': target,
    Authorization: `AWS4-HMAC-SHA256 Credential=${process.env.AMAZON_ACCESS_KEY}/${scope}, SignedHeaders=${signedHeaderNames}, Signature=${signature}`,
  };
}

export async function search(keyword, { limit = 10 } = {}) {
  const payload = JSON.stringify({
    Keywords: keyword,
    SearchIndex: 'Beauty',
    ItemCount: Math.min(limit, limits.itemsPerCall),
    PartnerTag: process.env.AMAZON_ASSOC_TAG,
    PartnerType: 'Associates',
    Marketplace: process.env.AMAZON_MARKETPLACE || 'www.amazon.com',
    Resources: ['ItemInfo.Title', 'ItemInfo.ByLineInfo', 'Offers.Listings.Price', 'Images.Primary.Medium', 'CustomerReviews.StarRating'],
  });
  const target = 'com.amazon.paapi5.v1.ProductAdvertisingAPIv1.SearchItems';
  const res = await fetch(`https://${HOST}${PATH}`, { method: 'POST', headers: signedHeaders(payload, target), body: payload });
  const text = await res.text();
  if (!res.ok) throw Object.assign(new Error(`Amazon PA-API ${res.status}: ${text.slice(0, 200)}`), { status: res.status });
  const json = JSON.parse(text);

  return (json?.SearchResult?.Items || []).map((it) => ({
    externalId: it.ASIN,
    name: it.ItemInfo?.Title?.DisplayValue || '',
    brand: it.ItemInfo?.ByLineInfo?.Brand?.DisplayValue || '',
    price: Math.round((it.Offers?.Listings?.[0]?.Price?.Amount || 0) * (Number(process.env.USD_KRW) || 1380)),
    url: it.DetailPageURL,
    image: it.Images?.Primary?.Medium?.URL,
    rating: it.CustomerReviews?.StarRating?.Value ?? null,
    reviews: 0,
    stock: Boolean(it.Offers?.Listings?.length),
    shippingDays: 7,
  }));
}

export async function ping() {
  const items = await search('vitamin c serum', { limit: 1 });
  return { ok: true, sample: items[0]?.name ?? null };
}
